### PR TITLE
Quantization Compressor Support 

### PR DESCRIPTION
**Pull request by Satrat**
Requires this compressed-tensors branch: https://github.com/neuralmagic/compressed-tensors/pull/45

* Adds support for saving compressed quantized models within SparseAutoModel saving. Compression type can be passed in via `quantization_format` or inferred from the model itself
* Simplified a lot of the save/load logic by moving it to helper classes in compressed-tensors

### Examples

Very little UX change, similar to sparsity we just pass `save_compressed=True` to enable compression. By default, we save weights in the fake_quant format is save_compressed isn't set.
```
from sparseml.transformers import SparseAutoModelForCausalLM, oneshot

recipe="tests/sparseml/transformers/compression/recipes/new_quant_full.yaml"
model_stub = "TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T"
dataset = "open_platypus"
max_seq_length = 512
num_calibration_samples = 512
output_dir = "./test_updated_llama1.1b_quant_compressed"

model = SparseAutoModelForCausalLM.from_pretrained(model_stub, device_map="cuda:0")
oneshot(
    model=model,
    dataset=dataset,
    overwrite_output_dir=True,
    output_dir=output_dir,
    max_seq_length=max_seq_length,
    num_calibration_samples=num_calibration_samples,
    recipe=recipe,
    pad_to_max_length=False,
    save_compressed=True
)
```

Reloading a fake_quant model and then compressing it:
```
from sparseml.transformers import SparseAutoModelForCausalLM

output_dir_fake = "./test_updated_llama1.1b_quant"
output_dir_compressed = "./test_updated_llama1.1b_quant_compressed"

model_reloaded = SparseAutoModelForCausalLM.from_pretrained(output_dir_fake)
model_reloaded.save_pretrained(output_dir_compressed, save_compressed=True)
```

You can also specify a quantization compression format by name. Right now we only have support for unpacked int quantization, but as we add additional compression formats for quantization this becomes more relevant
```
from sparseml.transformers import SparseAutoModelForCausalLM

output_dir_fake = "./test_updated_llama1.1b_quant"
output_dir_compressed = "./test_updated_llama1.1b_quant_compressed"

model_reloaded = SparseAutoModelForCausalLM.from_pretrained(output_dir_fake)
model_reloaded.save_pretrained(output_dir_compressed, quantization_format="int_quantized")
```

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
What would happen in the following case:


```python
from sparseml.transformers import SparseAutoModelForCausalLM, oneshot

recipe="tests/sparseml/transformers/compression/recipes/old_quant_full.yaml"
model_stub = "TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T"
dataset = "open_platypus"
max_seq_length = 512
num_calibration_samples = 512
output_dir = "./test_updated_llama1.1b_quant_compressed"

model = SparseAutoModelForCausalLM.from_pretrained(model_stub, device_map="cuda:0")
oneshot(
    model=model,
    dataset=dataset,
    max_seq_length=max_seq_length,
    num_calibration_samples=num_calibration_samples,
    recipe=recipe,
    pad_to_max_length=False,
)

model.save_pretrained("./saved_model", save_compressed=True)
```

**Comment by robertgshaw2-neuralmagic** [comment]
I think this looks very good. I like how it comes through via the `save_pretrained` method, which is very HF-naive!

Thoughts
- I think we should have `save_compressed=True` be the default. With quantization, we are going to need SparseML or vLLM to be able to consume these models anyways, I do not really see an advantage to serializing to fakequant

**Comment by Satrat** [comment]
> What would happen in the following case?
> 
> ```python
> from sparseml.transformers import SparseAutoModelForCausalLM, oneshot
> 
> recipe="tests/sparseml/transformers/compression/recipes/old_quant_full.yaml"
> model_stub = "TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T"
> dataset = "open_platypus"
> max_seq_length = 512
> num_calibration_samples = 512
> output_dir = "./test_updated_llama1.1b_quant_compressed"
> 
> model = SparseAutoModelForCausalLM.from_pretrained(model_stub, device_map="cuda:0")
> oneshot(
>     model=model,
>     dataset=dataset,
>     max_seq_length=max_seq_length,
>     num_calibration_samples=num_calibration_samples,
>     recipe=recipe,
>     pad_to_max_length=False,
> )
> 
> model.save_pretrained(output_dir, save_compressed=True)
> ```

This will throw a warning that we can't compress a model thats been quantized in the old format

**Comment by Satrat** [comment]
> I think this looks very good. I like how it comes through via the `save_pretrained` method, which is very HF-naive!
> 
> Thoughts
> 
> * I think we should have `save_compressed=True` be the default. With quantization, we are going to need SparseML or vLLM to be able to consume these models anyways, I do not really see an advantage to serializing to fakequant

My hesitation with this is once we compress the model we lose information on the original weights. In fakequant we retain the original uncompressed weights and can continue to recalibrate or run GPTQ on the model. Once its compressed we can't do this

File: src/sparseml/transformers/compression/quantization_format.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+def infer_quantization_format(
+    model, quantization_format: Optional[str] = None, save_compressed: bool = False
+) -> str:
+    """
+    Infers a quantization format based on model state and compression args
+
+    :param model: model to check for quantization, if the model is not quantized no
+    quantization format is returned
+    :param quantization_format: user provided quantization format, supercedes any
+    inferred quantization format
```
Comment (dbogunowicz) [inline src/sparseml/transformers/compression/quantization_format.py:34]: nit: indents :)

**Review by dbogunowicz — DISMISSED** [review]
🥇

**Review by bfineran — APPROVED** [review]

**Review by dbogunowicz — DISMISSED** [review]

File: src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py [hunk 2]
```diff
@@ -172,27 +172,38 @@ def fasterprune(
...
+
+                            q = fake_quantize(
+                                q,
+                                scale[:, i],
+                                zero_point[:, i],
+                                self.layer.quantization_scheme.weights,
+                            )
+
+                while q.ndim > 1:
                     q.squeeze()
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py:207]:
```suggestion
                    q = q.squeeze()
```

**Review by rahul-tuli — COMMENTED** [review]
LGTM after proposed fix

**Review by dbogunowicz — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Pull request merged**